deepinv/physics/unmixing.py

```diff
@@ -0,0 +1,186 @@
+
+    """
+
+    def __init__(
+        self, M: torch.Tensor = None, E: int = 15, C: int = 64, device="cpu", **kwargs
+    ):
+        super(HyperSpectralUnmixing, self).__init__()
+        self.device = device
+
+        if M is None:
+            # Simulate random normalised M
+            self.E, self.C = E, C
+            M = torch.rand((E, C), dtype=torch.float32)
+            M /= M.sum(dim=0, keepdim=True) * sqrt(self.C / self.E)
+        else:
+            self.E, self.C = M.shape
+
+        self.update_parameters(M=M, **kwargs)
+        self.M_pinv = torch.linalg.pinv(self.M)
+
+    def A(self, x: torch.Tensor, M: torch.Tensor = None, **kwargs):
+        r"""
+        Applies the endmembers matrix to the input abundances x.
+
+        :param torch.Tensor x: input abundances.
+        :param torch.Tensor M: optional new endmembers matrix :math:`\mathbf{M}` to be applied to the input abundances.
+        """
+        self.update_parameters(M=M, **kwargs)
+
+        if x.shape[1] != self.E:
+            raise ValueError("Number of endmembers in x should be as defined.")
+
+        return torch.einsum("ec,behw->bchw", self.M, x)
+
+    def A_dagger(self, y: torch.Tensor, M: torch.Tensor = None, **kwargs):
+        r"""
+        Applies the pseudoinverse endmember matrix to the image y.
+
+        :param torch.Tensor y: input image.
+        :param torch.Tensor M: optional new endmembers matrix :math:`\mathbf{M}` to be applied to the input image.
+        """
+        self.update_parameters(M=M, **kwargs)
+
+        if y.shape[1] != self.C:
+            raise ValueError("Number of channels in y should be as defined.")
+
+        return torch.einsum("ce,bchw->behw", self.M_pinv, y)
+
+    def A_adjoint(self, y: torch.Tensor, M: torch.Tensor = None, **kwargs):
+        r"""
+        Applies the transpose endmember matrix to the image y.
+
+        :param torch.Tensor y: input image.
+        :param torch.Tensor M: optional new endmembers matrix :math:`\mathbf{M}` to be applied to the input image.
+        """
+        self.update_parameters(M=M, **kwargs)
+
+        if y.shape[1] != self.C:
+            raise ValueError("Number of channels in y should be as defined.")
+
+        return torch.einsum("ce,bchw->behw", self.M.t(), y)
+
+    def update_parameters(self, M: torch.Tensor = None, **kwargs):
+        r"""
+        Updates the current endmembers matrix.
+
+        :param torch.Tensor M: New endmembers matrix to be applied to the input abundances.
+        """
+        if M is not None:
+            if M.shape != (self.E, self.C):
+                raise ValueError(
+                    "Number of endmembers and bands should be same as before."
+                )
+            self.M = torch.nn.Parameter(M.to(self.device), requires_grad=False)
+
+        if hasattr(self.noise_model, "update_parameters"):
+            self.noise_model.update_parameters(**kwargs)
+
+
+if __name__ == "__main__":
```

Comment:
I would remove this test debug code or comment it, since it will lower the codecoverage

Reply:
done